In [70]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler


In [3]:
# loading the dataset
df = pd.read_csv("/content/drive/MyDrive/Hamoye_Data/energydata_complete.csv", encoding= "latin-1")

In [4]:
# Exploring the dataset
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [5]:
df.tail()

date  Appliances  lights         T1       RH_1  \
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3    T4  ...    T9     RH_9  \
19730  25.890000  42.025714  27.200000  41.163333  24.7  ...  23.2  46.7900   
19731  25.754000  42.080000  27.133333  41.223333  24.7  ...  23.2  46.7900   
19732  25.628571  42.768571  27.050000  41.690000  24.7  ...  23.2  46.7900   
19733  25.414000  43.036000  26.890000  41.290000  24.7  ...  23.2  46.8175   
19734  25.264286  42.971429  26.823333  41.156667  24.7  ...  23.2  46.8450   

           T_out  Press_mm_hg     RH_out  Windspeed  Visibility  Tdewpoint  \
19730  22.733333        755.2  55.666667   3.333333   23.666667  13.333333   
19731  22.600000        755.2  56.000000   3.500000   24.500000  13.300000   
19732  22.466667        755.2  56.333333   3.666667   25.333333  13.266667   
19733  22.333333        755.2  56.666667   3.833333   26.166667  13.233333   
19734  22.200000        755.2  57.000000   4.000000   27.000000  13.200000   

             rv1        rv2  
19730  43.096812  43.096812  
19731  49.282940  49.282940  
19732  29.199117  29.199117  
19733   6.322784   6.322784  
19734  34.118851  34.118851  

[5 rows x 29 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [7]:
df.shape

(19735, 29)

In [8]:
df.isnull().sum()

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64

QUESTION 17

From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?

In [17]:
# Select the independent and dependent variables
X = df[['T2']]  # Temperature in the living room (independent variable)
y = df['T6']    # Temperature outside the building (dependent variable)

In [13]:
# Initialize and fit the linear regression model
model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [18]:
# Make predictions
y_pred = model.predict(X)

In [19]:
y_pred

array([ 5.37165597,  5.37165597,  5.37165597, ..., 19.67562202,
       19.19818742, 18.86506395])

In [20]:
# Calculate the Root Mean Squared Error (RMSE)
rmse = np.sqrt(mean_squared_error(y, y_pred))

ANSWER TO QUESTION 17

In [21]:
rmse

3.6444644873484573

QUESTION 18

Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:

What is the Mean Absolute Error (in three decimal places) for the  training set?

In [22]:
# Removing some columns
df = df.drop(columns=["date", "lights"])

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 27 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Appliances   19735 non-null  int64  
 1   T1           19735 non-null  float64
 2   RH_1         19735 non-null  float64
 3   T2           19735 non-null  float64
 4   RH_2         19735 non-null  float64
 5   T3           19735 non-null  float64
 6   RH_3         19735 non-null  float64
 7   T4           19735 non-null  float64
 8   RH_4         19735 non-null  float64
 9   T5           19735 non-null  float64
 10  RH_5         19735 non-null  float64
 11  T6           19735 non-null  float64
 12  RH_6         19735 non-null  float64
 13  T7           19735 non-null  float64
 14  RH_7         19735 non-null  float64
 15  T8           19735 non-null  float64
 16  RH_8         19735 non-null  float64
 17  T9           19735 non-null  float64
 18  RH_9         19735 non-null  float64
 19  T_ou

In [24]:
# Defining the features (independent variables) and the target variable
X = df.drop(columns=["Appliances"])
y = df["Appliances"]

In [27]:
# Split the dataset into training and testing sets with a 70-30 split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [28]:
# Initialize the MinMaxScaler
scaler = MinMaxScaler()

In [29]:
# Normalize the training and testing sets
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)

In [30]:
# Initialize and fit the multiple linear regression model
model = LinearRegression()
model.fit(X_train_normalized, y_train)

LinearRegression()

In [31]:
# Make predictions on the training set
y_train_pred = model.predict(X_train_normalized)

In [32]:
y_train_pred

array([ 49.97775425, 100.31632974,  40.12912536, ...,  59.47229745,
       103.23225103,  88.1415215 ])

In [36]:
# Calculate the Mean Absolute Error (MAE) for the training set
mae_train = mean_absolute_error(y_train, y_train_pred)

ANSWER TO QUESTION 18

In [50]:
print("MAE_train => ", round(mae_train, 3))

MAE_train =>  53.742


QUESTION 19

What is the Root Mean Squared Error (in three decimal places) for the training set?

In [49]:
# Calculating the root mean squared error (RMSE) for the training set
rmse_train = np.sqrt(mse_train)
print("RMSE_train => ", round(rmse_train, 3))

RMSE_train =>  95.216


QUESTION 20

What is the Mean Absolute Error (in three decimal places) for test set?

In [57]:
# Make predictions on the test set
y_test_pred = model.predict(X_test_normalized)

In [61]:
# Calculate the Mean Absolute Error (MAE) for the test set
mae_test = mean_absolute_error(y_test, y_test_pred)
print("MAE_test => ", round(mae_test, 3))

MAE_test =>  53.643


QUESTION 21

What is the Root Mean Squared Error (in three decimal places) for test set?

In [62]:
# Calculating the mean squared error (MSE) for the test set
mse_test = mean_squared_error(y_test, y_test_pred)

In [63]:
# Calculating the Root Mean Squared Error (RMSE) for the test set
rmse_test = np.sqrt(mse_test)
print("RMSE_test => ", round(rmse_test, 3))

RMSE_test =>  93.64


QUESTION 22

Did the Model above overfit to the training set



ANSWER TO 22

No

QUESTION 23

Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?



In [65]:
# Train a Ridge regression model with default parameters
ridge_model = Ridge()
ridge_model.fit(X_train_normalized, y_train)

Ridge()

In [66]:
# Make predictions on the test set
y_test_pred_ridge = ridge_model.predict(X_test_normalized)

In [68]:
# Calculating the Root Mean Squared Error (RMSE) for the test set with Ridge regression
rmse_test_ridge = np.sqrt(mean_squared_error(y_test, y_test_pred_ridge))
print("RMSE_test_ridge => ", round(rmse_test_ridge, 3))

RMSE_test_ridge =>  93.709


There is a very slight change in RMSE. RMSE_test =>  93.64 and RMSE_test_ridge =>  93.709



QUESTION 24

Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?



In [71]:
# Training a Lasso regression model with default parameters
lasso_model = Lasso()
lasso_model.fit(X_train_normalized, y_train)

Lasso()

In [72]:
# Get the feature weights from the trained Lasso regression model
lasso_weights = lasso_model.coef_

In [73]:
# Count the number of features with non-zero feature weights
non_zero_features = sum(lasso_weights != 0)


In [74]:
print("Number of non zero features is: ", non_zero_features)

Number of non zero features is:  4


QUESTION 25

What is the new RMSE with the Lasso Regression on the test set?



In [75]:
# Make predictions on the test set using the trained Lasso regression model
y_test_pred_lasso = lasso_model.predict(X_test_normalized)

In [79]:
# Calculate the RMSE for the test set with Lasso regression
rmse_test_lasso = np.sqrt(mean_squared_error(y_test, y_test_pred_lasso))
print("Lasso RMSE test is: ", round(rmse_test_lasso, 3))

Lasso RMSE test is:  99.424
